### PR TITLE
RSDK-4747 Limit capture frequency for RPLiDAR

### PR DESCRIPTION
**Pull request by jeremyrhyde**
**Summary**
This PR puts limits on the possible configured capture rate based on the rplidar model being used.

**Testing**
`capture_frequency_hz: -1`
```
2024-02-05T16:09:27.773Z	DEBUG	robot_server	impl/resource_manager.go:598	now configuring resource	{"resource":"rdk:component:camera/rplidar"}
2024-02-05T11:09:27.774-0500	INFO	robot_server.process.rplidar-module_/usr/local/bin/rplidar-module.StdOut	pexec/managed_process.go:244
\_ 2024-02-05T16:09:27.774Z	INFO	rplidarModule	rplidar/rplidar.go:122	attempting to connect to device at serial_path: /dev/tty.usbserial-0001
2024-02-05T11:09:33.825-0500	INFO	robot_server.process.rplidar-module_/usr/local/bin/rplidar-module.StdOut	pexec/managed_process.go:244
\_ 2024-02-05T16:09:33.825Z	INFO	rplidarModule	rplidar/rplidar.go:130	found and connected to an A1 rplidar
2024-02-05T16:09:33.826Z	ERROR	robot_server.rdk:component:camera/rplidar	resource/graph_node.go:230	resource build error: rpc error: code = Unknown desc = zero or negative capture frequency	{"resource":"rdk:component:camera/rplidar","model":"viam:lidar:rplidar"}
```

`capture_frequency_hz: 0`
```
2024-02-05T16:09:02.447Z	DEBUG	robot_server	impl/resource_manager.go:598	now configuring resource	{"resource":"rdk:component:camera/rplidar"}
2024-02-05T11:09:02.448-0500	INFO	robot_server.process.rplidar-module_/usr/local/bin/rplidar-module.StdOut	pexec/managed_process.go:244
\_ 2024-02-05T16:09:02.448Z	INFO	rplidarModule	rplidar/rplidar.go:122	attempting to connect to device at serial_path: /dev/tty.usbserial-0001
2024-02-05T11:09:08.495-0500	INFO	robot_server.process.rplidar-module_/usr/local/bin/rplidar-module.StdOut	pexec/managed_process.go:244
\_ 2024-02-05T16:09:08.495Z	INFO	rplidarModule	rplidar/rplidar.go:130	found and connected to an A1 rplidar
2024-02-05T16:09:08.496Z	ERROR	robot_server.rdk:component:camera/rplidar	resource/graph_node.go:230	resource build error: rpc error: code = Unknown desc = zero or negative capture frequency	{"resource":"rdk:component:camera/rplidar","model":"viam:lidar:rplidar"}
```

`capture_frequency_hz: 5`
```
2024-02-05T16:08:00.964Z	DEBUG	robot_server	impl/resource_manager.go:598	now configuring resource	{"resource":"rdk:component:camera/rplidar"}
2024-02-05T11:08:00.965-0500	INFO	robot_server.process.rplidar-module_/usr/local/bin/rplidar-module.StdOut	pexec/managed_process.go:244
\_ 2024-02-05T16:08:00.965Z	INFO	rplidarModule	rplidar/rplidar.go:122	attempting to connect to device at serial_path: /dev/tty.usbserial-0001
2024-02-05T11:08:07.054-0500	INFO	robot_server.process.rplidar-module_/usr/local/bin/rplidar-module.StdOut	pexec/managed_process.go:244
\_ 2024-02-05T16:08:07.054Z	INFO	rplidarModule	rplidar/rplidar.go:130	found and connected to an A1 rplidar
2024-02-05T11:08:07.055-0500	INFO	robot_server.process.rplidar-module_/usr/local/bin/rplidar-module.StdOut	pexec/managed_process.go:244
\_ 2024-02-05T16:08:07.054Z	DEBUG	rplidarModule	rplidar/rplidar.go:180	starting motor
```

`capture_frequency_hz: 20`
```
2024-02-05T16:08:29.751Z	DEBUG	robot_server	impl/resource_manager.go:598	now configuring resource	{"resource":"rdk:component:camera/rplidar"}
2024-02-05T11:08:29.753-0500	INFO	robot_server.process.rplidar-module_/usr/local/bin/rplidar-module.StdOut	pexec/managed_process.go:244
\_ 2024-02-05T16:08:29.753Z	INFO	rplidarModule	rplidar/rplidar.go:122	attempting to connect to device at serial_path: /dev/tty.usbserial-0001
2024-02-05T11:08:35.803-0500	INFO	robot_server.process.rplidar-module_/usr/local/bin/rplidar-module.StdOut	pexec/managed_process.go:244
\_ 2024-02-05T16:08:35.803Z	INFO	rplidarModule	rplidar/rplidar.go:130	found and connected to an A1 rplidar
2024-02-05T16:08:35.804Z	ERROR	robot_server.rdk:component:camera/rplidar	resource/graph_node.go:230	resource build error: rpc error: code = Unknown desc = set capture frequency (20) is greater than max frequency (10) for rplidar A1	{"resource":"rdk:component:camera/rplidar","model":"viam:lidar:rplidar"}
```

**JIRA Ticket:** https://viam.atlassian.net/browse/RSDK-4747

### CODE REVIEW

File: rplidar.go [hunk 1]
```diff
@@ -51,6 +51,8 @@ var (
 	Model = resource.NewModel("viam", "lidar", "rplidar")
 	// rplidarModelByteMap maps the byte model representation to a string representation
 	rplidarModelByteMap = map[byte]string{24: "A1", 49: "A3", 97: "S1"}
+	// The max capture frequency for rplidar models based on datasheets
```
Comment (kkufieta) [inline rplidar.go:54]: [opt] 
```suggestion
	// The max capture frequency for rplidar models is based on their datasheets
```

File: rplidar.go [hunk 2]
```diff
@@ -51,6 +51,8 @@ var (
...
+	maxScanningFrequencyByModel = map[string]float64{"A1": 10, "A3": 15, "S1": 15}
```
Comment (kkufieta) [inline rplidar.go:55]: [opt/nit] Since we're now using the models in multiple places, I'm in favor of making them constants

File: rplidar.go [hunk 3]
```diff
@@ -416,3 +433,28 @@ func getRplidarProcesses() ([]int, error) {
...
 }
+
+// getCaptureFrequencyHzFromConfig extract the capture_frequency_hz from the rplidar resource config
+func getCaptureFrequencyHzFromConfig(c resource.Config) (float64, error) {
+	var captureFreqHz float64
+	var captureMethodFound bool
+
+	for _, assocResourceCfg := range c.AssociatedResourceConfigs {
+		captureMethodsMapInterface := assocResourceCfg.Attributes["capture_methods"]
+		if captureMethodsMapInterface != nil {
```
Comment (kkufieta) [inline rplidar.go:444]: [nit]
```suggestion
		if captureMethodsMapInterface := assocResourceCfg.Attributes["capture_methods"]; captureMethodsMapInterface != nil {
```

**Review by kkufieta — APPROVED** [review]
LGTM % nits

**Review by JohnN193 — APPROVED** [review]

**Pull request merged**